### ISSUE TITLE
Automatically step over code with no source

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#245]
This would be like the "Just My Code" feature in Visual Studio, when performing a Step Into operation, have the debugger automatically step over functions without local source available.  This would probably require a change to the protocol where Duktape sends a REQ to the client (do you have source for this?) and if it gets a "no" reply back, do a Step Over instead.

The reason for this is that, currently, I have an issue where certain files (game maps notably) have code embedded in them and so the client is unable to locate the code for them by filename, leading to file not found exceptions.  So if I could get Duktape to treat these like external code instead of trying to step inside, that would be a great workaround.

The other alternative, of course, would be to have Duktape send the original source over the wire... But that would probably require invasive changes. :wink:

**Comment by svaarala** [comment]
That's an interesting idea, I'll have to think about the changes needed :)

Regarding sending the original source, that's actually been on the debugger todo list for a while. It'd be nice because one could then debug dynamic eval() code for example. It wouldn't be too bad to implement but there are some corner cases to it (some code is created using `Function ()` calls so they'll need a fake `function (...) { }` wrapper etc).

**Comment by dzzie** [comment]
this should be easy to do on the client side. I implemented some auto stepping to keep the users out of my library code. If the file name doesnt match the users file just automatically issue a step out. An automatic alternative, if all of your library code was added to the context with an eval() in script (not api), then I think the engine will auto step over it even if they issued a step in.

**Comment by fatcerberus** [comment]
Yeah, the Step Out trick I thought of last night, it's a very elegant solution. :)  No idea why I didn't think of it originally...

**Comment by fatcerberus** [comment]
@svaarala
I'm considering implementing the source tracking on my own, it would be quite useful for the new command-line debugger for minisphere I'm working on--such a tool would have a harder time locating sources than the IDE, having no context with which to locate the original project files.

In theory it would probably be enough to store the source code string as an internal property of the Function object, and then implement a GetSourceCode debugger command to return that string to the client.  Or am I completely off-base and it would actually be more complicated than that?

**Comment by svaarala** [comment]
That's roughly the approach I'd use too. There are some minor issues; e.g. not all functions have a full explicit source - you can create a function as e.g. `new Function('x', 'y', 'return x+y')` in which case the source property would not be provided by the user.

Because the source is rarely needed it'd be nice to be able to compress it somehow, but embedding a compression algorithm just for that purpose seems an overkill.

**Comment by fatcerberus** [comment]
There's a public-domain implementation of zlib floating around called miniz.  It's a single .c file, not sure how much it would add in terms of footprint.

**Comment by svaarala** [comment]
Yeah there are many alternatives like LZ4. But I'd like to avoid depending on a library as a mandatory dependency because it complicates porting, so source compression would probably need to be an option.

Anyway, source compression can be implemented separately, it's not really a core feature.

**Comment by fatcerberus** [comment]
I definitely agree there.  Source tracking itself should also be optional, as it increases memory pressure if every function object has to carry a source code string.

**Comment by svaarala** [comment]
Most definitely, keeping the source is not an option in many environments. I think it should be turned off by default.

**Comment by fatcerberus** [comment]
One thing I notice is that filename and line number information is tracked at the level of individual functions.  I realized that the same approach won't work here, as a single source file can contain multiple functions, and it's not much use if the debugger can only provide source for the currently executing function--even less so in the case of closures.

So, my initial thought was to add these two lines at the bottom of `duk_js_compile()`:
```c
	duk_push_lstring(ctx, (const char *)src_buffer, src_length);
	duk_put_prop_string(ctx, -2, "\xff" "Source");
```

But I'm thinking that won't work.  Supposing we have a source file that looks like this:
```js
function foo() {
  debugger;
  print("bar");
}

foo();
```

When Duktape pauses on the `debugger` statement, any attempt to retrieve the source code will try to read it from `foo._Source`--which of course will fail because the source text is associated with the "wrapper function" created by the compiler and not `foo` itself.  Is there a way to walk the scope chain?

**Comment by fatcerberus** [comment]
@svaarala Any tips for my dilemma above?

**Comment by fatcerberus** [comment]
I haven't had any luck with implementing this. :(  I'm assuming I have to associate the source code string with the closure returned here:
https://github.com/svaarala/duktape/blob/master/src/duk_api_compile.c#L132-L135

The trouble is that I don't know how to find that closure later given only an activation record.
